### DISCUSSION TITLE
Good testing practices with 🦔 Angular Testing Library

### DISCUSSION THREAD

**Discussion by giscus** [timdeschryver/timdeschryver.dev#280] [Announcements]
# Good testing practices with 🦔 Angular Testing Library

The Angular Testing Library provides utility functions to interact with Angular components, in the same way as a user would. This brings more maintainability to our tests, gives us more confidence that our component does what it's supposed to do, and it improves the accessibility which is better for our users. All these benefits, plus you'll see that it's fun to write tests in this way.

https://timdeschryver.dev/blog/good-testing-practices-with-angular-testing-library

**Comment by timsar2** [comment]
From angular 17, We have web test runner.
How to integrate this library with that?

**Reply by timdeschryver** [reply]
Are you experiencing any problems with it @timsar2 ?
I haven't tried it, but Testing Library is test runner agnostic, so it should just work.

**Comment by dougal83** [comment]
I had a bit of trouble with the `userEvent.type` not working in the forEach loop:
```typescript
inputValues.forEach(({ value, label }) => {
    const control = screen.getByLabelText(label);
    if (control.tagName === 'MAT-SELECT') {
      userEvent.selectOptions(control, value.toString());
    } else {
      userEvent.type(control, value.toString());
    }
  });
```
The following worked:
```typescript
for await (const { value, placeholder } of inputValues) {
    const control = screen.getByLabelText(label);
    if (control.tagName === 'MAT-SELECT') {
      userEvent.selectOptions(control, value.toString());
    } else {
      await userEvent.type(control, value.toString());
    }
  });
```
Would there be any reason for my difficulties withing an async test?  Thanks for these articles, they are great!

**Reply by timdeschryver** [reply]
Good question @dougal83 the time of writing this API was using user-event v13, which didn't have the async API.
With user-event v14 all events need to be awaited.

Using the `async`/`await` syntax might be harder to do when using `forEach`, for these cases I also fallback to using a regular `for` loop.

**Comment by NechiK** [comment]
Is there a way to somehow document each test case? I mean still use one big test, but still get list in the end like if this test was split by `it` or `describe` calls?
I've a really big form wizards with many steps and I wanna test the whole form with all possible flows, for example, click Next button, then got Back and check if state was saved (I don't know if that's correct approach).

**Reply by timdeschryver** [reply]
Depending on your test runner, you can use soft assertions for this case.
For example, this can be done with Playwright as mentioned in https://timdeschryver.dev/blog/improve-your-test-experience-with-playwright-soft-assertions

**Reply by NechiK** [reply]
I think you got it a little bit wrong. I want to reach the next behavior.

I want to render, for example, a big form wizard component (it has many steps with form elements each). And I want to run some happy path tests, navigation tests, etc. But as the output of such a test, I want to see some human-readable test cases.

Very simple test example:
```
test('Big Form Wizard, () => {
  expect(screen.getByText(/Next/i).closest('button').disabled).toEqual(true);

  // call some method that will add a message to output if the above test passed
  // For example:
  testCase('Next button is disabled at start');

  const firstNameInput = screen.getByLabelText(/First Name/i);
  const lastNameInput = screen.getByLabelText(/Last Name/i);
  await user.type('Maks');
  await user.type('N');

  expect(screen.getByText(/Next/i).closest('button').disabled).toEqual(false);

  // Call test case method again
  testCase('Next button is enabled when First and Last name entered');
});
```

So, in this case, I could have an output report with passed test cases. I thought I could just put component rendering into `beforeAll` and then test each part using `it`, but it looks like, in this case, Jest doesn't keep the Angular app state from previous `it` methods.

**Reply by timdeschryver** [reply]
Oh I see @NechiK 
You can set the `ATL_SKIP_AUTO_CLEANUP` environment variable to `true`. 
This will prevent a cleanup after every `it`.

**Reply by NechiK** [reply]
@timdeschryver It looks like it doesn't help because when I run `screen.debug();` on second `it` - it returns just `<body />`. If I run `screen.debug();` for the first `it` - I see the `<app-root>` tag inside the body and other parts/markup of my application.
And I also couldn't find any documentation for the `ATL_SKIP_AUTO_CLEANUP` env variable, only analogs for React and other frameworks.

Here's how I try to run it `ATL_SKIP_AUTO_CLEANUP=true npm run test.`
`npm run test` - is simple `ng test`;

UPD: I've found `ATL_SKIP_AUTO_CLEANUP` in one of the commits in the release, but still, it looks like the document body is cleaned up after each test.

UPD2: I've put console logs to the place where cleanup should be called and it wasn't called. However, the body is empty in each test (except the first one).

**Reply by NechiK** [reply]
One more observation. If I set `removeAngularAttributes: true`, then Testing Library keeps the Angular app in the body, and some tests pass, but it looks like change detection stopped working. Because when I call `await user.type(emailInput, someValue)` it doesn't update my Submit button disabled state.

Currently, I just implemented my custom collector class and methods to get all passed test cases, but I'd prefer to use "native" methods of Jest instead.

**Reply by timdeschryver** [reply]
@NechiK Oh I forgot Angular added this behavior in one of the latest versions.
By default Angular will teardown the TestBed and the rendered components after each test.
To opt-out of this behavior change the behavior of the TestBed using the [ModuleTeardownOptions](https://angular.io/api/core/testing/ModuleTeardownOptions), and set the `destroyAfterEach` property to `false`.

For example:

```ts
getTestBed().initTestEnvironment(
    BrowserDynamicTestingModule,
    platformBrowserDynamicTesting([]),
  {
    teardown: {
      destroyAfterEach: false
    }
  }
);
```

But it seems `ATL_SKIP_AUTO_CLEANUP=false` doesn't property clean up correctly, but that's a different issue 😅.

**Reply by NechiK** [reply]
@timdeschryver Thanks for pointing to the teardown options. I didn't know about it.

I decided to use Angular Testing Library API and configure it with the `configureTestBed` property. Am I right that we can make many `configureTestingModule` calls, and all settings put there will only overwrite the configuration (don't reset it with the provided config)?

```
configureTestBed: (testBed): void => {
        testBed.configureTestingModule(
          {
            teardown: { destroyAfterEach: false },
          },
        );
      },
```

However, it looks like the routing stopped working after I did this. I'll try to prepare some minimal examples, but here's how my test is working now:

1. Rendering `<app-root></app-root>` component (it has router-outlet inside)
2. In my code, the initial router is 'login' and uses SignInComponent.
3. Checking that the `Submit` button is disabled _**(now it's broken here because the router outlet doesn't render the SignIn Component)**_
4. Finding `Email` and `Password` fields, enter data, and click the `Submit` button
5. After it checks if I'm on the Dashboard page by searching its elements
6. Testing dashboard page

**Reply by timdeschryver** [reply]
Yea, that should work.
I didn't know this had impact on the routing.
If you have a reproduction feel free to create an issue at https://github.com/testing-library/angular-testing-library/issues

**Reply by NechiK** [reply]
Hi @timdeschryver 
I finally got time to create a reproduction repo and attached it to [the issue](https://github.com/testing-library/angular-testing-library/issues/444)

**Comment by idembele70** [comment]
Thank you !
